File: Baekjoon/Class3/09.2579 계단 오르기.js
const inputFile = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(inputFile, "utf8").trim().split("\n").map(Number);

const stepsN = input[0];

const maxScores = new Array(stepsN + 1).fill(0); // 타뷸레이션
if(stepsN >= 1) maxScores[1] = input[1];
if(stepsN >= 2) maxScores[2] = input[1] + input[2];
for(let i = 3; i <= stepsN; i++) {
    const leapBefore = input[i] + maxScores[i - 2];
    const noLeapBefore = input[i] + input[i - 1] + maxScores[i - 3];
    maxScores[i] = Math.max(leapBefore, noLeapBefore);
}
console.log(maxScores[stepsN]);

File: Baekjoon/Class3/09.2579 계단 오르기.py
import sys
input = sys.stdin.readline

steps_n = int(input())
steps = [int(input()) for _ in range(steps_n)]

max_scores = [0] * steps_n #타뷸레이션
if steps_n >= 1:
    max_scores[0] = steps[0]
if steps_n >= 2:
    max_scores[1] = steps[0] + steps[1]
if steps_n >= 3:
    max_scores[2] = max(steps[0] + steps[2], steps[1] + steps[2])
for i in range(3, steps_n):
    leap_before = steps[i] + max_scores[i - 2]
    no_leap_before = steps[i] + steps[i - 1] + max_scores[i - 3]
    max_scores[i] = max(leap_before, no_leap_before)
print(max_scores[steps_n - 1])